### PR TITLE
[Inductor] Allow output-input aliasing in torch.cond branches (#186104)

### PR DESCRIPTION
**Pull request by rmiao**
Summary:

Allow output-input aliasing in `torch.cond` branches.

torch.cond branches are mutually exclusive — only one branch executes at runtime. When the true branch returns an operand directly (e.g., identity/skip path), the output aliases the input. This is safe because the non-taken branch never runs, so there is no concurrent access to aliased buffers.

Previously, two checks rejected this pattern:
1. **Dynamo** (`higher_order_ops.py`): `CondHigherOrderVariable.supports_aliasing = False` rejected aliasing during tracing with a graph break.
2. **Functionalization** (`cond.py`): `_check_alias_and_mutation` rejected aliasing during AOT functionalization.

Note: Dynamo already allowed aliasing in inference (`supports_aliasing = not torch.is_grad_enabled()`), but `torch.export` traces with grad enabled even for inference models, so the check was ineffective for the AOTI export path.

Fix:
- Set `supports_aliasing = True` unconditionally on `CondHigherOrderVariable`. Aliasing is safe for cond regardless of grad context because branches are exclusive — even with autograd, only the taken branch's gradient flows.
- Skip the alias check in the functionalization impl (keep mutation check).

Motivation: PrismNet's `_clone_layer_outputs` uses `lo.clone()` to break output-input aliasing required by torch.cond. Removing this constraint saves ~0.046ms per inference (2 DHEN layers at B=2048).

Adds `test_cond_output_input_aliasing` regression test.

Test Plan:
```
buck2 test fbcode//mode/opt -c 'fbcode.nvcc_arch=a100,h100' fbcode//caffe2/test/inductor:control_flow -- --exact 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_device_cpu (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_device_cuda (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_multi_output_device_cpu (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_multi_output_device_cuda (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_both_branches_device_cpu (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_output_input_aliasing_both_branches_device_cuda (caffe2.test.inductor.test_control_flow.CondTests)' 'fbcode//caffe2/test/inductor:control_flow - test_cond_input_mutation_still_rejected (caffe2.test.inductor.test_control_flow.CondTests)'
```
Test session: https://www.internalfb.com/intern/testinfra/testrun/10414574313882076

PrismNet repro — all 11 tests pass, including `CondOutputInputAlias: without clone` which previously failed with "Replace return input with return input.clone()". Confirms `lo.clone()` workaround in `_clone_layer_outputs` can be removed.
```
buck2 run fbcode//mode/opt fbcode//aps_models/ads/gmp/models/ads_mtml_prism_net_dedicated_model/experimental/benchmarks:repro_aoti_inductor_bugs
```

Reviewed By: kqfu, atalman

Differential Revision: D106106380

@diff-train-skip-merge

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by kqfu** [comment]
topic: not user facing

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -2363,7 +2363,9 @@ class CondHigherOrderVariable(TorchHigherOrderOperatorVariable):
     _HOP_NAME = "torch.cond"
     _ALLOW_FALLBACK_TO_EAGER = False
     supports_input_mutation = False
-    supports_aliasing = False
+    # torch.cond branches are mutually exclusive — only one branch executes,
+    # so output-input aliasing is safe (no concurrent access to aliased buffers).
+    supports_aliasing = True
```
Comment (kqfu) [inline torch/_dynamo/variables/higher_order_ops.py:2370]: Curious before this diff, does it crash or does it just add extra clones in subgraph?

File: torch/_higher_order_ops/cond.py [hunk 2]
```diff
@@ -732,10 +735,15 @@ def cond_func(ctx, pred, true_fn, false_fn, inputs):
         functional_true = ctx.functionalize(_maybe_run_with_interpreter(true_fn))
         functional_false = ctx.functionalize(_maybe_run_with_interpreter(false_fn))
         pre_dispatch = hasattr(ctx, "mode") and ctx.mode.pre_dispatch
+        # torch.cond branches are mutually exclusive — only one executes at
+        # runtime, so output-input aliasing is safe.  Skip the alias check
+        # but still check for input mutation.
         for branch, branch_name in [(true_fn, "cond_true"), (false_fn, "cond_false")]:
-            _check_alias_and_mutation(
-                branch, unwrapped_inputs, branch_name, pre_dispatch
+            _, inp_mutation = has_potential_input_alias_or_mutation(
```
Comment (kqfu) [inline torch/_higher_order_ops/cond.py]: Also curious why mutation is not allowed. Actually in [torch.cond document](https://docs.pytorch.org/docs/2.12/generated/torch.cond.html), it says `The function can perform in-place mutations on its input tensors during inference`

**Comment by zou3519** [comment]
@claude review these changes

**Review by zou3519 — CHANGES_REQUESTED** [review]
requesting changes for discussion. I haven't read the PR yet, but my stance on this problem is that we need a lot of functionalization machinery to do output-input aliasing and I see no functionalization changes

**Comment by rmiao** [comment]
> requesting changes for discussion. I haven't read the PR yet, but my stance on this problem is that we need a lot of functionalization machinery to do output-input aliasing and I see no functionalization changes

Good point. Fixed in cond_func by cloning branch outputs that alias an input (reusing the existing clone_outputs_aliasing_inputs helper), keeping mutation rejected. So the cond HOP stays functional.

**Review by kqfu — APPROVED** [review]

**Comment by rmiao** [comment]
> This PR has pending changes requested. Please address the comments and update the PR before merging.

Already addressed the comments. the only 1 CI was trunk issue.

**Review by zou3519 — CHANGES_REQUESTED** [review]
discussed on the reland PR
